File: src/principle/open_close.ts
/// 例子: 人物换肤

/**
 * @description 抽象类
 */
abstract class AbstractSkin {
    public abstract display(): void
}

/**
 * @description 默认皮肤
 */
class DefaultSkin extends AbstractSkin {
    public display(): void {
        console.log('using default skin.')
    }
}

/**
 * @description 自定义皮肤1
 */
class MySkin1 extends AbstractSkin {
    public display() {
        console.log('using my skin 1.')
    }
}

/**
 * @description 人物
 */
class Player {
    private skin: AbstractSkin | null = null;

    public setSkin(skin: AbstractSkin) {
        this.skin = skin
    }

    public display() {
        if (!this.skin) console.log('no skin yet!')
        else this.skin.display()
    }
}


// test
(() => {
    const defaultSkin = new DefaultSkin()
    const mySkin1 = new MySkin1()
    const player = new Player()

    player.display()
// no skin yet!

    player.setSkin(defaultSkin)
    player.display()
// using default skin.

    player.setSkin(mySkin1)
    player.display()
// using my skin 1.
})()

File: src/principle/open_close.rs
use std::ops::Deref;

/// region abstract skin
pub trait AbstractSkin {
    fn display(&self) -> ();
}
// endregion

// region skin - default
pub struct DefaultSkin {}

impl AbstractSkin for DefaultSkin {
    fn display(&self) {
        println!("using default skin.");
    }
}
// endregion

// region skin - 1
pub struct MySkin1 {}

impl AbstractSkin for MySkin1 {
    fn display(&self) {
        println!("using my skin 1.");
    }
}
// endregion

// region player
pub struct Player {
    skin: Box<dyn AbstractSkin>,
}

impl Player {
    pub fn new(skin: Box<dyn AbstractSkin>) -> Player {
        Player { skin }
    }

    pub fn set_skin(&mut self, skin: Box<dyn AbstractSkin>) {
        self.skin = skin;
    }

    fn display(&self) {
        self.skin.deref().display();
    }
}
// endregion

pub fn open_close_example() {
    let default_skin = DefaultSkin {};
    let skin1 = MySkin1 {};

    let mut player1 = Player::new(Box::new(default_skin));

    player1.display();
    // using default skin.

    player1.set_skin(Box::new(skin1));
    player1.display();
    // using my skin 1.
}
